src/session/session_op_tracker.c

```diff
@@ -0,0 +1,252 @@
+/*-
+ * Copyright (c) 2014-2015 MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+/*
+ * __wt_session_op_tracker_create_entry --
+ *	Setup an entry in the operation tracker queue.
+ */
+int
+__wt_session_op_tracker_create_entry(
+    WT_SESSION_IMPL *session, uint32_t type,
+    uint32_t api_boundary, WT_OP_TRACKER_ENTRY **entryp)
+{
+	WT_OP_TRACKER_ENTRY *entry, *prev_entry;
+
+	*entryp = NULL;
+	if (F_ISSET(session, WT_SESSION_INTERNAL))
+		return (0);
+
+	/* Don't capture tracing unless we are recording for the API */
+	if (!api_boundary && TAILQ_EMPTY(&session->op_trackerq))
+		return (0);
+
+	/*
+	 * Don't capture WT_CONNECTION methods. It would be nice to have
+	 * a different way to capture this. An alternative would be to have
+	 * an exclude list built via op_track.py. In the mean time this
+	 * works.
+	 */
+	if (type >= WT_OP_TYPE_WT_CONNECTION_async_flush &&
+	    type <= WT_OP_TYPE_WT_CONNECTION_get_extension_api)
+		return (0);
+
+	WT_ASSERT(session, session->api_call_depth == 0 ||
+	    !TAILQ_EMPTY(&session->op_trackerq));
+
+	/*
+	 * If we are entering the first time via a public API clear any
+	 * tracked operations from the previous API call. Track the call depth
+	 * since WiredTiger uses API calls internally.
+	 */
+	if (api_boundary && session->api_call_depth == 0)
+		WT_RET(__wt_session_op_tracker_clear(session));
+
+	if ((entry = TAILQ_FIRST(&session->op_tracker_availq)) != NULL) {
+		TAILQ_REMOVE(&session->op_tracker_availq, entry, aq);
+	} else
+		WT_RET(__wt_calloc_one(session, &entry));
+
+	WT_RET(__wt_epoch(session, &entry->start));
+	memcpy(&entry->last_stop, &entry->start, sizeof(struct timespec));
+	entry->api_boundary = api_boundary;
+	TAILQ_INSERT_TAIL(&session->op_trackerq, entry, q);
+
+	/*
+	 * Find the parent for this entry. It is the last entry in the queue
+	 * that hasn't been finished.
+	 */
+	for (prev_entry = TAILQ_PREV(entry, __op_tracker, q);
+	    prev_entry != NULL;
+	    prev_entry = TAILQ_PREV(prev_entry, __op_tracker, q)) {
+		if (prev_entry->done)
+			continue;
+		prev_entry->self_time_us +=
+		    WT_TIMEDIFF(entry->start, prev_entry->last_stop);
+		break;
+	}
+
+	/* Create and insert the first entry into the slow op tracker. */
+	entry->type = type;
+	*entryp = entry;
+
+	if (api_boundary)
```

Comment:
The ```api_boundary``` is a flag passed into ```__wt_session_op_tracker_create_entry```. The ```session->api_call_depth``` variable is used to track how deep in the API call stack we are.

We need to track those since WiredTiger makes calls via it's own APIs, and we want to track when we are at the top level (user) API call.